pkg/dbmate/clickhouse.go

```diff
@@ -0,0 +1,258 @@
+	}
+	defer mustClose(db)
+
+	_, err = db.Exec("create database " + clickhouseQuoteIdentifier(name))
+
+	return err
+}
+
+// DropDatabase drops the specified database (if it exists)
+func (drv ClickHouseDriver) DropDatabase(u *url.URL) error {
+	name := drv.databaseName(u)
+	fmt.Printf("Dropping: %s\n", name)
+
+	db, err := drv.openClickHouseDB(u)
+	if err != nil {
+		return err
+	}
+	defer mustClose(db)
+
+	_, err = db.Exec("drop database if exists " + clickhouseQuoteIdentifier(name))
+
+	return err
+}
+
+func clickhouseSchemaDump(db *sql.DB, buf *bytes.Buffer, databaseName string) error {
+	buf.WriteString("\n--\n-- Database schema\n--\n\n")
+
+	buf.WriteString("CREATE DATABASE " + clickhouseQuoteIdentifier(databaseName) + " IF NOT EXISTS;\n\n")
+
+	tables, err := queryColumn(db, "show tables")
+	if err != nil {
+		return err
+	}
+	sort.Strings(tables)
+
+	for _, table := range tables {
+		var clause string
+		err = db.QueryRow("show create table " + clickhouseQuoteIdentifier(table)).Scan(&clause)
+		if err != nil {
+			return err
+		}
+		buf.WriteString(clause + "\n\n")
+	}
+	return nil
+}
+
+func clickhouseSchemaMigrationsDump(db *sql.DB, buf *bytes.Buffer) error {
+	// load applied migrations
+	migrations, err := queryColumn(db,
+		"select version from schema_migrations final where applied order by version asc",
+	)
+	if err != nil {
+		return err
+	}
+
+	quoter := strings.NewReplacer(`\`, `\\`, `'`, `\'`)
+	for i := range migrations {
+		migrations[i] = "'" + quoter.Replace(migrations[i]) + "'"
+	}
+
+	// build schema_migrations table data
+	buf.WriteString("\n--\n-- Dbmate schema migrations\n--\n\n")
+
+	if len(migrations) > 0 {
+		buf.WriteString("INSERT INTO schema_migrations (version) VALUES\n    (" +
+			strings.Join(migrations, "),\n    (") +
+			");\n")
+	}
+
+	return nil
+}
+
+// DumpSchema returns the current database schema
+func (drv ClickHouseDriver) DumpSchema(u *url.URL, db *sql.DB) ([]byte, error) {
+	var buf bytes.Buffer
+	var err error
+
+	err = clickhouseSchemaDump(db, &buf, drv.databaseName(u))
+	if err != nil {
+		return nil, err
+	}
+
+	err = clickhouseSchemaMigrationsDump(db, &buf)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// DatabaseExists determines whether the database exists
+func (drv ClickHouseDriver) DatabaseExists(u *url.URL) (bool, error) {
+	name := drv.databaseName(u)
+
+	db, err := drv.openClickHouseDB(u)
+	if err != nil {
+		return false, err
+	}
+	defer mustClose(db)
+
+	exists := false
+	err = db.QueryRow("SELECT 1 FROM system.databases where name = ?", name).
+		Scan(&exists)
+	if err == sql.ErrNoRows {
+		return false, nil
+	}
+
+	return exists, err
+}
+
+// CreateMigrationsTable creates the schema_migrations table
+func (drv ClickHouseDriver) CreateMigrationsTable(db *sql.DB) error {
+	_, err := db.Exec(`
+		create table if not exists schema_migrations (
+			version String,
+			ts DateTime default now(),
+			applied UInt8 default 1
+		) engine = ReplacingMergeTree(ts)
```

Comment:
can you keep this consistent with the table we create for other drivers. we might add an applied timestamp at some point, but I'd like to be consistent and do it for all drivers.

Reply:
The reason this column (and a ts column) exist is that ClickHouse is append-only DB. There is no usable way to update or delete a row. So, in order to be able to mark a migration as rolled back, we need to insert a new row with the same version but applied = false and newer ts. This is an idiomatic way to do updates in ClickHouse.

Reply:
Makes sense, thanks